[SIL] Make owned function arguments lexical.

nate-chandler opened the pull request:
The destroys of owned arguments must not be hoisted over deinit barriers to respect the semantics of lexical lifetimes.  Here, owned SILValues which are SILFunctionArguments are made to be lexical.  Code which hoists destroys will check this field and not hoist destroys.

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test

atrick commented:
As discussed, we should fix ownership RAUW before doing this so most of the CSE/SILCombine optimizations will still work

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test

nate-chandler committed:
[RAUW] Allow RAUWing some lexical values.

RAUWing a lexical value with a non-lexical value is illegal because it
would result in the value's lifetime being shortened without regard to
deinit barriers.  RAUWing _with_ a lexical value is LEGAL so long as
doing so doesn't extend its lifetime.

nate-chandler committed:
[LexicalDestroyHoisting] Eliminate updateSSA.

UpdateSSA was being used to rewrite destroys, not for its intended
purpose.  Besides being a misuse, it's also wasteful.  Instead, just
rewrite the destroys that it would have updated.

nate-chandler committed:
[SIL] Make owned function arguments lexical.

The destroys of owned arguments must not be hoisted over deinit barriers
to respect the semantics of lexical lifetimes.  Here, owned SILValues
which are SILFunctionArguments are made to be lexical.  Code which
hoists destroys will check this field and not hoist destroys.

Allowed more RAUWing to be done with lexical values; specifically,
allowed the replacement of a non-lexical value with a lexical value if
the replacing value's lifetime wouldn't be extended as a result.

Updated tests to get owned values via applies rather than having them
passed as arguments.

nate-chandler commented:
@swift-ci please benchmark

nate-chandler committed:
[LexicalDestroyHoisting] Mark context const.

nate-chandler commented:
@swift-ci please test

eeckstein commented:
@nate-chandler Can you please document that in SIL.rst? E.g. under "Function Types" and/or "Ownership SSA".
It would be nice to have a paragraph explaining lexical lifetimes in general (e.g. under "Ownership SSA").

nate-chandler commented:
@eeckstein Absolutely!  Sorry for leaving that out.

nate-chandler committed:
[Gardening] Tweaked comment.

nate-chandler commented:
Added some documentation to SIL.rst.

nate-chandler commented:
@swift-ci please test

eeckstein commented on docs/SIL.rst:
```diff
@@ -2364,6 +2364,62 @@ The current list of interior pointer SIL instructions are:
+
+1: `begin_borrow [lexical]`
+2: `move_value [lexical]`
+3: @owned function arguments
```
`alloc_stack [lexical]`?

atrick replied:
You could point out that `alloc_stack [lexical]` constrains the lifetime of an addressable variable. Since the variable has an in-memory representation, no lexical SILValue is required.

nate-chandler replied:
Added.

eeckstein reviewed (commented):
The documentation in SIL.rst looks very good!

Do you know the cause of the benchmark regressions?

atrick commented on docs/SIL.rst:
```diff
@@ -2364,6 +2364,62 @@ The current list of interior pointer SIL instructions are:
+   class `D`, if `d` is weakly referenced via a local variable, loading from
+   that weak reference and the deallocating `c` will result in `d` remaining
+   alive.  Hoisting the destroy of `c` above the load of `d`, however, would
+   result in `d` being deallocated too.
```
Should be something like:

    Given an instance `c` of a class `C` which weakly references an instance `d` of a second
    class `D`, if `d` is referenced via a local variable `v`, then loading
    that weak reference from `c` within the variable scope should return a non-nil reference to `d`.
    Hoisting the destroy of `v` above the weak load from `c`, however, would result in the destruction of
    `d` before that load and a nil weak reference to `D`.

nate-chandler replied:
Fixed.

nate-chandler committed:
[SIL] Doc'd deinit barriers.

nate-chandler commented on lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:
```diff
@@ -349,24 +384,44 @@ bool OwnershipRAUWHelper::hasValidRAUWOwnership(SILValue oldValue,
   // uses for liveness extension.
-  return findExtendedTransitiveGuaranteedUses(oldValue,
-                                              context.guaranteedUsePoints);
+  if (!findOldValueUses(oldValue, context.guaranteedUsePoints))
```
Fixed.

nate-chandler commented on lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:
```diff
@@ -349,24 +384,44 @@ bool OwnershipRAUWHelper::hasValidRAUWOwnership(SILValue oldValue,
   return true;
 }
 
+static bool findOldValueUses(SILValue oldValue,
```
Updated.

nate-chandler replied:
Updated again.  Added a new find uses of an owned value including borrows non-reborrowing scope ending uses `findUsesOfSimpleOwnedValue`.  Added check to make sure we don't bail out in non-lexical cases.

nate-chandler commented on lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:
```diff
@@ -293,17 +293,52 @@ static void cleanupOperandsBeforeDeletion(SILInstruction *oldValue,
-                                                SILValue newValue) {
+bool OwnershipRAUWHelper::hasValidRAUWOwnership(
+    SILValue oldValue, SILValue newValue,
+    SmallVectorImpl<Operand *> const &oldUses) {
```
Fixed.

nate-chandler commented on lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:
```diff
@@ -293,17 +293,52 @@ static void cleanupOperandsBeforeDeletion(SILInstruction *oldValue,
 //                      Ownership RAUW Helper Functions
 //===----------------------------------------------------------------------===//
 
+bool OwnershipRAUWHelper::oldValueUsesAreWithinNewLexicalValueLiveness(
```
Fixed.

nate-chandler commented on docs/SIL.rst:
```diff
@@ -2364,6 +2364,62 @@ The current list of interior pointer SIL instructions are:
+3. Given an instance of a class which owns a buffer and deallocates it on
+   deinitialization, writing to the pointer and then deallocating the instance
+   is defined behavior.  Hoisting the destroy of the instance above the write to
+   the memory would result in writing to a freed pointer.
```
It's supposed to be a concrete example, but fixed.

nate-chandler committed:
[OwnershipUtils] Added find for simple value.

nate-chandler committed:
[LexicalDestroyFolding] Used ArrayRef.

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

atrick commented on lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:
```diff
@@ -353,20 +368,53 @@ bool OwnershipRAUWHelper::hasValidRAUWOwnership(SILValue oldValue,
+                                                      ownedUsePoints);
+  }
+  default:
+    return OwnershipRAUWHelper::hasValidRAUWOwnership(oldValue, newValue, {});
```
> default:
>     return OwnershipRAUWHelper::hasValidRAUWOwnership(oldValue, newValue, {});

I think the same lifetime extension happens when an unowned value's uses is replaced with a guaranteed or owned value. So it should probably be just like the owned case here.

nate-chandler replied:
Fixed.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test windows platform

atrick reviewed (approved):
Great!

nate-chandler commented:
The root cause of the `Set.subtracting.Box.Empty` benchmark regression is that we don't yet have the lexical destroy hoisting utility to run during copy propagation on owned lexical values (`move_value [lexical]` and `@owned` arguments).  That results in extra ARC traffic in `$sSh11subtractingyShyxGABF`.

In detail, the problem is that CanonicalizeOSSALifetime, which doesn't respect lexical lifetimes, no longer hoists the destroys of `%1` of

```
sil [serialized] [ossa] @$sSh11subtractingyShyxGABF : $@convention(method) <Element where Element : Hashable> (@guaranteed Set<Element>, @owned Set<Element>) -> @owned Set<Element> {
bb0(%0 : @guaranteed $Set<Element>, %1 : @owned $Set<Element>):
```

to just after its last uses.  Specifically, the destroy is not hoisted up from the end of the function to just after the apply of `$sSh12_subtractingyShyxGqd__7ElementQyd__RszSTRd__lF`.

Before this change, SILCombine saw

```
  %102 = copy_value %1 : $Set<Element>
  %105 = function_ref @$sSh12_subtractingyShyxGqd__7ElementQyd__RszSTRd__lF : $@convention(method) <τ_0_0 where τ_0_0 : Hashable><τ_1_0 where τ_0_0 == τ_1_0.Element, τ_1_0 : Sequence> (@in_guaranteed τ_1_0, @owned Set<τ_0_0>) -> @owned Set<τ_0_0>
  %106 = apply %105<Element, Set<Element>>(%103, %102) : $@convention(method) <τ_0_0 where τ_0_0 : Hashable><τ_1_0 where τ_0_0 == τ_1_0.Element, τ_1_0 : Sequence> (@in_guaranteed τ_1_0, @owned Set<τ_0_0>) -> @owned Set<τ_0_0>
  destroy_value %1 : $Set<Element>
```

and was able to rewrite it as

```
  %85 = function_ref @$sSh12_subtractingyShyxGqd__7ElementQyd__RszSTRd__lF : $@convention(method) <τ_0_0 where τ_0_0 : Hashable><τ_1_0 where τ_0_0 == τ_1_0.Element, τ_1_0 : Sequence> (@in_guaranteed τ_1_0, @owned Set<τ_0_0>) -> @owned Set<τ_0_0>
  %86 = apply %85<Element, Set<Element>>(%83, %1) : $@convention(method) <τ_0_0 where τ_0_0 : Hashable><τ_1_0 where τ_0_0 == τ_1_0.Element, τ_1_0 : Sequence> (@in_guaranteed τ_1_0, @owned Set<τ_0_0>) -> @owned Set<τ_0_0>
```

Now, though, the `destroy_value` isn't in the same block, so the extra copy can't be eliminated.

eeckstein commented:
Is this something which is easy to fix?

nate-chandler commented:
It should be reasonably straightforward.  ShrinkBorrowScope has been restructed ( https://github.com/apple/swift/pull/41358 ) to make it easy to create the new LexicalDestroyHoisting from it.  Filed rdar://88903962 .
